hpilo_boot: fix module failing when trying to power on an already powered-on server

ryanb74 opened the pull request:
##### SUMMARY
For this module to be idempotent, it should be successful when trying to reach a power state which is already there.

It was already the case for "poweroff", however when running the module with "boot_once" state, it was failing if the server was already powered_on, which is not an idempotent behavior

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
hpilo_boot.py

##### ADDITIONAL INFORMATION

Before change, when running the module with state = "boot_once", the module fails :
```
Tuesday 28 January 2025  11:08:58 +0100 (0:00:00.012)       0:00:00.344 *******
fatal: [localhost]: FAILED! => {
    "changed": false,
    "msg": "HP iLO (server) reports that the server is already powered on !"
}

PLAY RECAP *******************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0  
```

After change,  when running the module with state = "boot_once", the module is successful and doesn't change anything : 
```
Tuesday 28 January 2025  11:11:51 +0100 (0:00:00.012)       0:00:00.326 *******
ok: [localhost] => {
    "changed": false,
    "power": "ON"
}

PLAY RECAP *******************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

russoz commented on plugins/modules/hpilo_boot.py:
```diff
@@ -188,9 +188,8 @@ def main():
-            module.fail_json(msg='HP iLO (%s) reports that the server is already powered on !' % host)
-
-        if power_status == 'ON':
+           pass
+        elif power_status == 'ON':
```
Hi @ryanb74 Thanks for your contribution!

Please beware this is a breaking change - people might have code out there that _expects_ it to fail. We need to take the long route on this one - add the new behaviour with a feature flag, deprecate the "old" behaviour, and later on remove the "old" behaviour entirely (along with the feature flag).

ryanb74 replied:
Hi @russoz, thanks for the review

Sorry for the late answer, I don't have much time to work on this.

I found [this markown ](https://github.com/russoz-ansible/ansible-contrib-unofficial/blob/main/deprecations.md) from one of your repositories, I'll try to fix the code according to it when possible and then update this issue

felixfontein replied:
Please do not yet deprecate the old behavior. First add the new option, the deprecation should add later so users have a chance to adjust before the deprecation warning hits them.

felixfontein commented:
@ryanb74 ping

needs_info

ryanb74 commented:
Hi,
I changed the code to allow the new behavior by setting the new parameter : `idempotent_boot_once` to true.

This setting defaults to false in which case the old behavior still keeps going.
I also added a deprecation warning, however I'm not sure which version of the community collection should be specified here so I just put 11.0.0

russoz commented on changelogs/fragments/9646-hpilo-fix-idempotency.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - "hpilo_boot - Add option to get an idempotent behavior while powering on server (success instead of failure when using 'state: boot_once' option) (https://github.com/ansible-collections/community.general/pull/9646)."
```
```suggestion
  - "hpilo_boot - Add option to get an idempotent behavior while powering on server, resulting success instead of failure when using ``state: boot_once`` option (https://github.com/ansible-collections/community.general/pull/9646)."
```

ryanb74 replied:
Thanks again for the review, I committed your proposal

russoz reviewed (commented):
Hi @ryanb74 Thanks for your contribution!

Just a minor nit on the changelog fragment, other than that LGTM

russoz commented on plugins/modules/hpilo_boot.py:
```diff
@@ -76,6 +76,11 @@
     choices: ["SSLv3", "SSLv23", "TLSv1", "TLSv1_1", "TLSv1_2"]
+  idempotent_boot_once:
+    description:
+      - "This makes the boot_once idempotent see github PR: https://github.com/ansible-collections/community.general/pull/9646"
```
Sorry, one more nit to be picked.
```suggestion
      - "This makes the boot_once idempotent see github PR: U(https://github.com/ansible-collections/community.general/pull/9646)."
```

felixfontein commented on plugins/modules/hpilo_boot.py:
```diff
@@ -78,7 +78,7 @@
   idempotent_boot_once:
     description:
-      - "This makes the boot_once idempotent see github PR: U(https://github.com/ansible-collections/community.general/pull/9646)."
+      - "This options makes boot_once succeed instead of failing when the server is already powered on."
```
```suggestion
      - "This options makes O(state=boot_once) succeed instead of failing when the server is already powered on."
```

ryanb74 replied:
```suggestion
      - "This option makes O(state=boot_once) succeed instead of failing when the server is already powered on."
```

felixfontein commented on plugins/modules/hpilo_boot.py:
```diff
@@ -76,6 +76,11 @@
+    description:
+      - "This option makes O(state=boot_once) succeed instead of failing when the server is already powered on."
+    type: bool
+    default: false
```
Sorry, it looks like we totally forgot to mention this:
```suggestion
    default: false
    version_added: 10.6.0
```
With that I think everything looks good!

ryanb74 replied:
Thanks for the review ! 👍

russoz reviewed (approved):
LGTM

felixfontein commented:
If you want you can create a PR that uncomments the deprecation (and adds a "deprecated_features" changelog fragment); that can be merged into `main` (and not backported to `stable-10`), so it will appear in community.general 11.0.0.
